lua/nvim-treesitter/textobjects.lua

```diff
@@ -66,13 +67,158 @@ function M.select_textobject(query_string)
     if smallest_range.start then
       local start_range = {smallest_range.start.node:range()}
       local node_range = {smallest_range.node:range()}
-      ts_utils.update_selection(bufnr, {start_range[1], start_range[2], node_range[3], node_range[4]})
+      return bufnr, {start_range[1], start_range[2], node_range[3], node_range[4]}, smallest_range.node
     else
-      ts_utils.update_selection(bufnr, smallest_range.node)
+      return bufnr, {smallest_range.node:range()}, smallest_range.node
     end
   end
 end
 
+function M.select_textobject(query_string)
+  local bufnr, textobject = textobject_at_point(query_string)
+  if textobject then
+    ts_utils.update_selection(bufnr, textobject)
+  end
+end
+
+local function swap_textobject(query_string, direction)
+  local bufnr, textobject_range, node = textobject_at_point(query_string)
+  if not node then return end
+
+  local step = direction > 0 and 1 or -1
+  local overlapping_range_ok = false
+  local same_parent = true
+  for _ = 1, math.abs(direction), step do
+    if direction > 0 then
+      ts_utils.swap_nodes(textobject_range,
+                          M.next_textobject(node, query_string, same_parent, overlapping_range_ok, bufnr),
+                          bufnr,
+                          "yes, set cursor!")
+    else
+      ts_utils.swap_nodes(textobject_range,
+                          M.previous_textobject(node, query_string, same_parent, overlapping_range_ok, bufnr),
+                          bufnr,
+                          "yes, set cursor!")
+    end
+  end
+end
+
+function M.swap_next(query_string)
+  swap_textobject(query_string, 1)
+end
+
+function M.swap_previous(query_string)
+  swap_textobject(query_string, -1)
+end
+
+function M.goto_adjacent(query_string, forward, start, same_parent, overlapping_range_ok)
+  local bufnr, _, node = textobject_at_point(query_string)
+  local adjacent_textobject
+  if forward then
+    adjacent_textobject = M.next_textobject(node,  query_string, same_parent, overlapping_range_ok, bufnr)
+  else
+    adjacent_textobject = M.previous_textobject(node,  query_string, same_parent, overlapping_range_ok, bufnr)
+  end
```

Comment:
i think you could do the same thing here, caching the function that takes the exact same arguments so you can do a 2 line function call instead of a full 6 lines if/else statement :)